Feature/sweepformula ap firing

t-b opened the pull request:
- [x] apfrequency tests
- [x] Documentation rst
- [x] Documentation notebook (coordinate with #345)
- [x] Add formulas to rst

Close #379.

timjarsky commented:
@t-b I was able to make it run. nice

default doesn't work; method is null

frequency values can be incorrect e.g., 10 e-3; they look like they are out by a factor of 1000

I'll upload a test file to your ftp that has the sweep browser set up

![image](https://user-images.githubusercontent.com/5693697/69474086-f2d89580-0d70-11ea-8acd-b61dc3422346.png)

timjarsky commented:
@t-b filename: apfrequency

t-b commented:
@timjarsky Should be fixed.

timjarsky commented:
@t-b `findlevels` and find level wrapper don't agree on the number of levels found in a wave.

```
•print dimsize(FindLevelWrapper(root:MIES:Analysis:slice_002_site_000_2019_11__0:ITC1600_Dev_0:sweep:X_155:AD_10, 0, FINDLEVEL_EDGE_INCREASING, FINDLEVEL_MODE_MULTI),cols)
  21
•findlevels/edge=1 root:MIES:Analysis:slice_002_site_000_2019_11__0:ITC1600_Dev_0:sweep:X_155:AD_10, 0
  V_Flag= 1; V_LevelsFound= 14; 
```

t-b commented:
@ukos-git Do you know why

```
	try
		WAVE output = SF_FormulaExecutor(SF_FormulaParser("apfrequency([1, 2])")); AbortOnRTE
		PASS()
	catch
		CHECK(0)
	endtry
```

fails?

t-b commented:
@timjarsky Thanks for the extra checking. I wanted to be really smart and reached the opposite end of the scale. Should be fixed now. And also with some short documentation and some more tests for `apfrequency`.

ukos-git commented:
apfrequency([1, 2]) Evaluates to apfrequency(1, 2) as function arguments are an array anyway. 

apfrequency([[1, 2]]) could work.

t-b commented:
@timjarsky I've also added a help entry for `vs` under plotting, externalized the sweepformula help  notebook for better maintanability, and changed the returned value for the instantaneous method to NaN if no levels could be found.

t-b commented:
Oh and changed to use your favourite markers by default

timjarsky commented:
@t-b it looks like `apfrequency` returns a the same value for all sweeps:

![image](https://user-images.githubusercontent.com/5693697/69765842-69d3ac80-112a-11ea-9787-8000848af132.png)

t-b commented:
@timjarsky Thanks, obvious bug with obvious fix.

timjarsky commented:
@t-b last pair of channels is being ignored by SF with AD/DA or AD#/DA#

![image](https://user-images.githubusercontent.com/5693697/69777175-edee5a00-1153-11ea-8039-48e45e625daf.png)

It's getting close! Thank you!!

timjarsky commented:
A couple of obvious bottlenecks;

Total time: 7.62961, Time in Function code: 1.78314 (23.4%)
Top function percentages:
Function MIES_SweepFormula.ipf SF_FormulaExecutor: 61%
Function MIES_SweepFormula.ipf SF_GetSweepForFormula: 32%

Annotated Top Functions:

*******************************************************************************************
Function: MIES_SweepFormula.ipf SF_FormulaExecutor; Percent total 61%
*******************************************************************************************
[00]          	|Function/WAVE SF_FormulaExecutor(jsonID, [jsonPath, graph])
[00]          	|	Variable jsonID
[00]          	|	String jsonPath
[00]          	|	String graph
[00]          	|
[00]          	|	Variable i, j, numIndices, JSONtype, mode
[00]          	|	string info
[00]          	|
[00]          	|	if(ParamIsDefault(jsonPath))
[00]          	|		jsonPath = ""
[00]          	|	endif
[00]          	|	if(ParamIsDefault(graph))
[00]          	|		graph = ""
[00]          	|	endif
[00]          	|
[00]          	|	// object and array evaluation
[00]          	|	JSONtype = JSON_GetType(jsonID, jsonPath)
[00]          	|	if(JSONtype == JSON_NUMERIC)
[00]          	|		Make/FREE out = { JSON_GetVariable(jsonID, jsonPath) }
[00]          	|		return out
[00]          	|	elseif(JSONtype == JSON_STRING)
[00]          	|		Make/FREE/T outT = { JSON_GetString(jsonID, jsonPath) }
[00]          	|		return outT
[00]          	|	elseif(JSONtype == JSON_ARRAY)
[00]          	|		WAVE topArraySize = JSON_GetMaxArraySize(jsonID, jsonPath)
[00]          	|		Make/FREE/N=(topArraySize[0])/B types = JSON_GetType(jsonID, jsonPath + "/" + num2str(p))
[00]          	|
[00]          	|		if(topArraySize[0] != 0 && types[0] == JSON_STRING)
[00]          	|			ASSERT(DimSize(topArraySize, ROWS) <= 1, "Text Waves Must Be 1-dimensional.")
[00]          	|			return JSON_GetTextWave(jsonID, jsonPath)
[00]          	|		endif
[00]          	|		EXTRACT/FREE types, strings, types[p] == JSON_STRING
[00]          	|		ASSERT(DimSize(strings, ROWS) == 0, "Object evaluation For Mixed Text/Numeric Arrays Is Not Allowed")
[00]          	|		WaveClear strings
[00]          	|
[00]          	|		ASSERT(DimSize(topArraySize, ROWS) < 4, "Unhandled Data Alignment. Only 3 Dimensions Are Supported For Operations.")
[00]          	|		WAVE out = JSON_GetWave(jsonID, jsonPath)
[00]          	|
[00]          	|		Redimension/N=4 topArraySize
[00]          	|		topArraySize[] = topArraySize[p] != 0 ? topArraySize[p] : 1
[00]          	|		Redimension/N=(topArraySize[0], topArraySize[1], topArraySize[2], topArraySize[3])/E=1 out
[00]          	|
[00]          	|		EXTRACT/FREE/INDX types, indices, (types[p] == JSON_OBJECT) || (types[p] == JSON_ARRAY)
[00]          	|		if(DimSize(indices, ROWS) == 1 && DimSize(out, ROWS) == 1)
[00]          	|			return SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/" + num2str(indices[0]), graph = graph)
[00]          	|		endif
[00]          	|		for(i = 0; i < DimSize(indices, ROWS); i += 1)
[00]          	|			WAVE element = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/" + num2str(indices[i]), graph = graph)
[00]          	|			if(DimSize(element, CHUNKS) > 1)
[00]          	|				DebugPrint("Merging Chunks To Layers for object: " + jsonPath + "/" + num2str(indices[i]))
[00]          	|				Redimension/N=(-1, -1, max(1, DimSize(element, LAYERS)) * DimSize(element, CHUNKS), 0)/E=1 element
[00]          	|			endif
[00]          	|			topArraySize[1,*] = max(topArraySize[p], DimSize(element, p - 1))
[00]          	|			if((DimSize(out, ROWS)   < topArraySize[0]) || \
[00]          	|			   (DimSize(out, COLS)   < topArraySize[1]) || \
[00]          	|			   (DimSize(out, LAYERS) < topArraySize[2]) || \
[00]          	|			   (DimSize(out, CHUNKS) < topArraySize[3]))
[03]*         	|				Redimension/N=(topArraySize[0], topArraySize[1], topArraySize[2], topArraySize[3]) out
[00]          	|			endif
[00]          	|			SF_FormulaWaveScaleTransfer(element, out, ROWS, COLS)
[00]          	|			SF_FormulaWaveScaleTransfer(element, out, COLS, LAYERS)
[00]          	|			SF_FormulaWaveScaleTransfer(element, out, LAYERS, CHUNKS)
[08]*         	|			out[indices[i]][0, max(0, DimSize(element, ROWS) - 1)][0, max(0, DimSize(element, COLS) - 1)][0, max(0, DimSize(element, LAYERS) - 1)] = element[q][r][s]
[00]          	|		endfor
[00]          	|
[00]          	|		EXTRACT/FREE/INDX types, indices, types[p] == JSON_NUMERIC
[00]          	|		for(i = 0; i < DimSize(indices, ROWS); i += 1)
[49]*****     	|			out[indices[i]][][][] = out[indices[i]][0][0][0]
[00]          	|		endfor
[00]          	|
[00]          	|		topArraySize[1,*] = topArraySize[p] == 1 ? 0 : topArraySize[p]
[00]          	|		Redimension/N=(topArraySize[0], topArraySize[1], topArraySize[2], topArraySize[3])/E=1 out
[00]          	|
[00]          	|		return out
[00]          	|	endif
[00]          	|
[00]          	|	// operation evaluation
[00]          	|	ASSERT(JSONtype == JSON_OBJECT, "Topmost element needs to be an object")
[00]          	|	WAVE/T operations = JSON_GetKeys(jsonID, jsonPath)
[00]          	|	ASSERT(DimSize(operations, ROWS) == 1, "Only one operation is allowed")
[00]          	|	jsonPath += "/" + SF_EscapeJsonPath(operations[0])
[00]          	|	ASSERT(JSON_GetType(jsonID, jsonPath) == JSON_ARRAY, "An array is required to hold the operands of the operation.")
[00]          	|	strswitch(operations[0])
[00]          	|		case "cursors":
[00]          	|		case "sweeps":
[00]          	|			WAVE/T wvT = JSON_GetTextWave(jsonID, jsonPath)
[00]          	|			break
[00]          	|		case "setscale":
[00]          	|		case "butterworth":
[00]          	|		case "channels":
[00]          	|		case "data":
[00]          	|		case "wave":
[00]          	|		case "findlevel":
[00]          	|			break
[00]          	|		default:
[00]          	|			WAVE wv = SF_FormulaExecutor(jsonID, jsonPath = jsonPath, graph = graph)
[00]          	|	endswitch
[00]          	|
[00]          	|	/// @name SweepFormulaOperations
[00]          	|	/// @{
[00]          	|	strswitch(LowerStr(operations[0]))
[00]          	|		case "-":
[00]          	|			if(DimSize(wv, ROWS) == 1)
[00]          	|				MatrixOP/FREE out = sumCols((-1) * wv)^t
[00]          	|			else
[00]          	|				MatrixOP/FREE out = (row(wv, 0) + sumCols((-1) * subRange(wv, 1, numRows(wv) - 1, 0, numCols(wv) - 1)))^t
[00]          	|			endif
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, COLS, ROWS)
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, LAYERS, COLS)
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, CHUNKS, LAYERS)
[00]          	|			Redimension/N=(-1, DimSize(out, LAYERS), DimSize(out, CHUNKS), 0)/E=1 out
[00]          	|			break
[00]          	|		case "+":
[00]          	|			MatrixOP/FREE out = sumCols(wv)^t
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, COLS, ROWS)
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, LAYERS, COLS)
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, CHUNKS, LAYERS)
[00]          	|			Redimension/N=(-1, DimSize(out, LAYERS), DimSize(out, CHUNKS), 0)/E=1 out
[00]          	|			break
[00]          	|		case "~1": // division
[00]          	|			ASSERT(DimSize(wv, ROWS) >= 2, "At least two operands are required")
[00]          	|			MatrixOP/FREE out = (row(wv, 0) / productCols(subRange(wv, 1, numRows(wv) - 1, 0, numCols(wv) - 1)))^t
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, COLS, ROWS)
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, LAYERS, COLS)
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, CHUNKS, LAYERS)
[00]          	|			Redimension/N=(-1, DimSize(out, LAYERS), DimSize(out, CHUNKS), 0)/E=1 out
[00]          	|			break
[00]          	|		case "*":
[00]          	|			MatrixOP/FREE out = productCols(wv)^t
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, COLS, ROWS)
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, LAYERS, COLS)
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, CHUNKS, LAYERS)
[00]          	|			Redimension/N=(-1, DimSize(out, LAYERS), DimSize(out, CHUNKS), 0)/E=1 out
[00]          	|			break
[00]          	|		case "range":
[00]          	|			/// range (start[, stop[, step]])
[00]          	|		case "…":
[00]          	|			ASSERT(DimSize(wv, CHUNKS) <= 1, "Unhandled dimension")
[00]          	|			ASSERT(DimSize(wv, LAYERS) <= 1, "Unhandled dimension")
[00]          	|			ASSERT(DimSize(wv, COLS) <= 1, "Unhandled dimension")
[00]          	|			if(DimSize(wv, ROWS) == 3)
[00]          	|				Make/N=(ceil(abs((wv[0] - wv[1]) / wv[2])))/FREE out = wv[0] + p * wv[2]
[00]          	|			elseif(DimSize(wv, ROWS) == 2)
[00]          	|				Make/N=(abs(trunc(wv[0])-trunc(wv[1])))/FREE out = wv[0] + p
[00]          	|			elseif(DimSize(wv, ROWS) == 1)
[00]          	|				Make/N=(abs(trunc(wv[0])))/FREE out = p
[00]          	|			else
[00]          	|				ASSERT(0, "Operation accepts 2-3 operands")
[00]          	|			endif
[00]          	|			break
[00]          	|		case "min":
[00]          	|			ASSERT(DimSize(wv, LAYERS) <= 1, "Unhandled dimension")
[00]          	|			ASSERT(DimSize(wv, CHUNKS) <= 1, "Unhandled dimension")
[00]          	|			MatrixOP/FREE out = minCols(wv)^t
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, COLS, ROWS)
[00]          	|			break
[00]          	|		case "max":
[00]          	|			ASSERT(DimSize(wv, LAYERS) <= 1, "Unhandled dimension")
[00]          	|			ASSERT(DimSize(wv, CHUNKS) <= 1, "Unhandled dimension")
[01]*         	|			MatrixOP/FREE out = maxCols(wv)^t
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, COLS, ROWS)
[00]          	|			break
[00]          	|		case "avg":
[00]          	|		case "mean":
[00]          	|			ASSERT(DimSize(wv, LAYERS) <= 1, "Unhandled dimension")
[00]          	|			ASSERT(DimSize(wv, CHUNKS) <= 1, "Unhandled dimension")
[00]          	|			MatrixOP/FREE out = averageCols(wv)^t
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, COLS, ROWS)
[00]          	|			break
[00]          	|		case "rms":
[00]          	|			ASSERT(DimSize(wv, LAYERS) <= 1, "Unhandled dimension")
[00]          	|			ASSERT(DimSize(wv, CHUNKS) <= 1, "Unhandled dimension")
[00]          	|			MatrixOP/FREE out = sqrt(averageCols(magsqr(wv)))^t
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, COLS, ROWS)
[00]          	|			break
[00]          	|		case "variance":
[00]          	|			ASSERT(DimSize(wv, LAYERS) <= 1, "Unhandled dimension")
[00]          	|			ASSERT(DimSize(wv, CHUNKS) <= 1, "Unhandled dimension")
[00]          	|			MatrixOP/FREE out = (sumCols(magSqr(wv - rowRepeat(averageCols(wv), numRows(wv))))/(numRows(wv) - 1))^t
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, COLS, ROWS)
[00]          	|			break
[00]          	|		case "stdev":
[00]          	|			ASSERT(DimSize(wv, LAYERS) <= 1, "Unhandled dimension")
[00]          	|			ASSERT(DimSize(wv, CHUNKS) <= 1, "Unhandled dimension")
[00]          	|			MatrixOP/FREE out = (sqrt(sumCols(powR(wv - rowRepeat(averageCols(wv), numRows(wv)), 2))/(numRows(wv) - 1)))^t
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, COLS, ROWS)
[00]          	|			break
[00]          	|		case "derivative":
[00]          	|			Make/FREE out
[00]          	|			ASSERT(DimSize(wv, ROWS) > 1, "Can not differentiate single point waves")
[00]          	|			Differentiate/DIM=(ROWS) wv/D=out
[00]          	|			CopyScales wv, out
[00]          	|			SetScale/P x, DimOffset(wv, ROWS), DimDelta(wv, ROWS), "d/dx", out
[00]          	|			break
[00]          	|		case "integrate":
[00]          	|			Make/FREE out
[00]          	|			ASSERT(DimSize(wv, ROWS) > 1, "Can not integrate single point waves")
[00]          	|			Integrate/METH=1/DIM=(ROWS) wv/D=out
[00]          	|			CopyScales wv, out
[00]          	|			SetScale/P x, DimOffset(wv, ROWS), DimDelta(wv, ROWS), "dx", out
[00]          	|			break
[00]          	|		case "butterworth":
[00]          	|			/// `butterworth(data, lowPassCutoff, highPassCutoff, order)`
[00]          	|			ASSERT(JSON_GetArraySize(jsonID, jsonPath) == 4, "The butterworth filter requires 4 arguments")
[00]          	|			WAVE data = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/0", graph = graph)
[00]          	|			WAVE lowPassCutoff = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/1")
[00]          	|			ASSERT(DimSize(lowPassCutoff, ROWS) == 1, "Too many input values for parameter lowPassCutoff")
[00]          	|			ASSERT(IsNumericWave(lowPassCutoff), "lowPassCutoff parameter must be numeric")
[00]          	|			WAVE highPassCutoff = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/2")
[00]          	|			ASSERT(DimSize(highPassCutoff, ROWS) == 1, "Too many input values for parameter highPassCutoff")
[00]          	|			ASSERT(IsNumericWave(highPassCutoff), "highPassCutoff parameter must be numeric")
[00]          	|			WAVE order = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/3")
[00]          	|			ASSERT(DimSize(order, ROWS) == 1, "Too many input values for parameter order")
[00]          	|			ASSERT(IsNumericWave(order), "order parameter must be numeric")
[00]          	|			FilterIIR/HI=(highPassCutoff[0] / WAVEBUILDER_MIN_SAMPINT_HZ)/LO=(lowPassCutoff[0] / WAVEBUILDER_MIN_SAMPINT_HZ)/ORD=(order[0])/DIM=(ROWS) data
[00]          	|			ASSERT(V_flag == 0, "FilterIIR returned error")
[00]          	|			WAVE out = data
[00]          	|			break
[00]          	|		case "time":
[00]          	|		case "xvalues":
[00]          	|			Make/FREE/N=(DimSize(wv, ROWS), DimSize(wv, COLS), DimSize(wv, LAYERS), DimSize(wv, CHUNKS)) out = DimOffset(wv, ROWS) + p * DimDelta(wv, ROWS)
[00]          	|			break
[00]          	|		case "setscale":
[00]          	|			/// `setscale(data, [dim, [dimOffset, [dimDelta[, unit]]]])
[00]          	|			numIndices = JSON_GetArraySize(jsonID, jsonPath)
[00]          	|			ASSERT(numIndices < 6, "Maximum number of arguments exceeded.")
[00]          	|			ASSERT(numIndices > 1, "At least two arguments.")
[00]          	|			WAVE data = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/0", graph = graph)
[00]          	|			WAVE/T dimension = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/1")
[00]          	|			ASSERT(DimSize(dimension, ROWS) == 1 && GrepString(dimension[0], "[x,y,z,t]") , "undefined input for dimension")
[00]          	|
[00]          	|			if(numIndices >= 3)
[00]          	|				WAVE offset = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/2")
[00]          	|				ASSERT(DimSize(offset, ROWS) == 1, "wrong usage of argument")
[00]          	|			else
[00]          	|				Make/FREE/N=1 offset  = {0}
[00]          	|			endif
[00]          	|			if(numIndices >= 4)
[00]          	|				WAVE delta = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/3")
[00]          	|				ASSERT(DimSize(delta, ROWS) == 1, "wrong usage of argument")
[00]          	|			else
[00]          	|				Make/FREE/N=1 delta = {1}
[00]          	|			endif
[00]          	|			if(numIndices == 5)
[00]          	|				WAVE/T unit = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/4")
[00]          	|				ASSERT(DimSize(unit, ROWS) == 1, "wrong usage of argument")
[00]          	|			else
[00]          	|				Make/FREE/N=1/T unit = {""}
[00]          	|			endif
[00]          	|
[00]          	|			strswitch(dimension[0])
[00]          	|				case "x":
[00]          	|					SetScale/P x, offset[0], delta[0], unit[0], data
[00]          	|					ASSERT(DimDelta(data, ROWS) == delta[0], "Encountered Igor Bug.")
[00]          	|					break
[00]          	|				case "y":
[00]          	|					SetScale/P y, offset[0], delta[0], unit[0], data
[00]          	|					ASSERT(DimDelta(data, COLS) == delta[0], "Encountered Igor Bug.")
[00]          	|					break
[00]          	|				case "z":
[00]          	|					SetScale/P z, offset[0], delta[0], unit[0], data
[00]          	|					ASSERT(DimDelta(data, LAYERS) == delta[0], "Encountered Igor Bug.")
[00]          	|					break
[00]          	|				case "t":
[00]          	|					SetScale/P t, offset[0], delta[0], unit[0], data
[00]          	|					ASSERT(DimDelta(data, CHUNKS) == delta[0], "Encountered Igor Bug.")
[00]          	|					break
[00]          	|			endswitch
[00]          	|			WAVE out = data
[00]          	|			break
[00]          	|		case "wave":
[00]          	|			ASSERT(JSON_GetArraySize(jsonID, jsonPath) == 1, "First argument is wave")
[00]          	|			WAVE/T wavelocation = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/0")
[00]          	|			WAVE out = $(wavelocation[0])
[00]          	|			break
[00]          	|		case "merge":
[00]          	|			ASSERT(DimSize(wv, LAYERS) <= 1, "Unhandled dimension")
[00]          	|			ASSERT(DimSize(wv, CHUNKS) <= 1, "Unhandled dimension")
[00]          	|			MatrixOP/FREE transposed = wv^T
[00]          	|			Extract/FREE transposed, out, (p < (JSON_GetType(jsonID, jsonPath + "/" + num2str(q)) != JSON_ARRAY ? 1 : JSON_GetArraySize(jsonID, jsonPath + "/" + num2str(q))))
[00]          	|			SetScale/P x, 0, 1, "", out
[00]          	|			break
[00]          	|		case "channels":
[00]          	|			/// `channels([str name]+)` converts a named channel from string to numbers.
[00]          	|			///
[00]          	|			/// returns [[channelName, channelNumber]+]
[00]          	|			String channelName, channelNumber
[00]          	|			String regExp = "^(?i)(" + ReplaceString(";", ITC_CHANNEL_NAMES, "|") + ")([0-9]+)?$"
[00]          	|			numIndices = JSON_GetArraySize(jsonID, jsonPath)
[00]          	|
[00]          	|			Make/N=(numIndices, 2)/FREE out = NaN
[00]          	|			for(i = 0; i < numIndices; i += 1)
[00]          	|				JSONtype = JSON_GetType(jsonID, jsonPath + "/" + num2str(i))
[00]          	|				channelName = ""
[00]          	|				if(JSONtype == JSON_NUMERIC)
[00]          	|					out[i][1] = JSON_GetVariable(jsonID, jsonPath + "/" + num2str(i))
[00]          	|				elseif(JSONtype == JSON_STRING)
[00]          	|					SplitString/E=regExp JSON_GetString(jsonID, jsonPath + "/" + num2str(i)), channelName, channelNumber
[00]          	|					if(V_flag == 0)
[00]          	|						continue
[00]          	|					endif
[00]          	|					out[i][1] = str2num(channelNumber)
[00]          	|				endif
[00]          	|				ASSERT(!isFinite(out[i][1]) || out[i][1] < NUM_MAX_CHANNELS, "Maximum Number Of Channels exceeded.")
[00]          	|				out[i][0] = WhichListItem(channelName, ITC_CHANNEL_NAMES, ";", 0, 0)
[00]          	|			endfor
[00]          	|			out[][] = out[p][q] < 0 ? NaN : out[p][q]
[00]          	|			break
[00]          	|		case "sweeps":
[00]          	|			/// `sweeps([str type])`
[00]          	|			///  @p type: `|displayed|all`
[00]          	|			///           displayed (default): get (selected) sweeps
[00]          	|			///           al:                  get all possible sweeps
[00]          	|			ASSERT(JSON_GetArraySize(jsonID, jsonPath) <= 1, "Function requires 1 argument at most.")
[00]          	|
[00]          	|			JSONtype = JSON_GetType(jsonID, jsonPath + "/0")
[00]          	|			if(JSONtype == JSON_NULL)
[00]          	|				wvT[0] = "displayed"
[00]          	|			endif
[00]          	|
[00]          	|			strswitch(wvT[0])
[00]          	|				case "all":
[00]          	|					mode = OVS_SWEEP_ALL_SWEEPNO
[00]          	|					break
[00]          	|				case "displayed":
[00]          	|					mode = OVS_SWEEP_SELECTION_SWEEPNO
[00]          	|					break
[00]          	|				default:
[00]          	|					ASSERT(0, "Undefined argument")
[00]          	|			endswitch
[00]          	|
[00]          	|			ASSERT(!ParamIsDefault(graph) && !IsEmpty(graph), "Graph not specified.")
[00]          	|			WAVE/Z out = OVS_GetSelectedSweeps(graph, mode)
[00]          	|			if(!WaveExists(out) && mode == OVS_SWEEP_SELECTION_SWEEPNO)
[00]          	|				WAVE/Z out = OVS_GetSelectedSweeps(graph, OVS_SWEEP_ALL_SWEEPNO)
[00]          	|			endif
[00]          	|			if(!WaveExists(out))
[00]          	|				Make/N=1/FREE out = {NaN} // simulates [null]
[00]          	|			endif
[00]          	|			break
[00]          	|		case "data":
[00]          	|			/// `data(array range,array channels,array sweeps)`
[00]          	|			///
[00]          	|			/// returns [[sweeps][channel]] for all [sweeps] in list sweepNumbers, grouped by channels
[00]          	|			ASSERT(!ParamIsDefault(graph) && !IsEmpty(graph), "Graph for extracting sweeps not specified.")
[00]          	|
[00]          	|			WAVE range = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/0", graph = graph)
[00]          	|			ASSERT(DimSize(range, ROWS) == 2, "A range can not hold more than two points.")
[00]          	|			range = !IsNaN(range[p]) ? range[p] : (p == 0 ? -1 : 1) * inf
[00]          	|
[00]          	|			WAVE channels = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/1")
[00]          	|			ASSERT(DimSize(channels, COLS) == 2, "A channel input consists of [[channelName, channelNumber]+].")
[00]          	|
[00]          	|			WAVE sweeps = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/2", graph = graph)
[00]          	|			ASSERT(DimSize(sweeps, COLS) < 2, "Sweeps are one-dimensional.")
[00]          	|
[00]*         	|			WAVE/Z out = SF_GetSweepForFormula(graph, range[0], range[1], channels[0][0], channels[0][1], sweeps)
[00]          	|			if(!WaveExists(out))
[00]          	|				DebugPrint("Call to SF_GetSweepForFormula returned no results")
[00]          	|				Make/FREE/N=1 out = {NaN}
[00]          	|				break
[00]          	|			endif
[00]          	|			numIndices = DimSize(channels, ROWS)
[00]          	|			Redimension/N=(-1, -1, numIndices) out
[00]          	|			j = 1
[00]          	|			for(i = 1; i < numIndices; i += 1)
[00]          	|				WAVE wv = SF_GetSweepForFormula(graph, range[0], range[1], channels[i][0], channels[i][1], sweeps)
[00]          	|				if(!WaveExists(wv))
[00]          	|					continue
[00]          	|				endif
[00]          	|				out[][][j] = wv[p][q]
[00]          	|				j += 1
[00]          	|				WaveClear wv
[00]          	|			endfor
[00]          	|			Redimension/N=(-1, -1, j) out
[00]          	|			break
[00]          	|		case "log": // JSON logic debug operation
[00]          	|			print wv[0]
[00]          	|			WAVE out = wv
[00]          	|			break
[00]          	|		case "log10": // decadic logarithm
[00]          	|			ASSERT(DimSize(wv, LAYERS) <= 1, "Unhandled dimension")
[00]          	|			ASSERT(DimSize(wv, CHUNKS) <= 1, "Unhandled dimension")
[00]          	|			MatrixOP/FREE out = log(wv)
[00]          	|			SF_FormulaWaveScaleTransfer(wv, out, COLS, ROWS)
[00]          	|			break
[00]          	|		case "cursors":
[00]          	|			Make/FREE/N=(DimSize(wvT, ROWS)) out = NaN
[00]          	|			for(i = 0; i < DimSize(wvT, ROWS); i += 1)
[00]          	|				ASSERT(GrepString(wvT[i], "^(?i)[A-J]$"), "Invalid Cursor Name")
[00]          	|				if(ParamIsDefault(graph))
[00]          	|					out[i] = xcsr($wvT[i])
[00]          	|				else
[00]          	|					info = CsrInfo($wvT[i], graph)
[00]          	|					if(IsEmpty(info))
[00]          	|						continue
[00]          	|					endif
[00]          	|					out[i] = xcsr($wvT[i], graph)
[00]          	|				endif
[00]          	|			endfor
[00]          	|			break
[00]          	|		case "findlevel":
[00]          	|			// findlevel(data, level, [edge])
[00]          	|			numIndices = JSON_GetArraySize(jsonID, jsonPath)
[00]          	|			ASSERT(numIndices <=3, "Maximum number of arguments exceeded.")
[00]          	|			ASSERT(numIndices > 1, "At least two arguments.")
[00]          	|			WAVE data = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/0", graph = graph)
[00]          	|			WAVE level = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/1")
[00]          	|			ASSERT(DimSize(level, ROWS) == 1, "Too many input values for parameter level")
[00]          	|			ASSERT(IsNumericWave(level), "level parameter must be numeric")
[00]          	|			if(numIndices == 3)
[00]          	|				WAVE edge = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/2")
[00]          	|				ASSERT(DimSize(edge, ROWS) == 1, "Too many input values for parameter level")
[00]          	|				ASSERT(IsNumericWave(edge), "level parameter must be numeric")
[00]          	|			else
[00]          	|				Make/FREE edge = {0}
[00]          	|			endif
[00]          	|
[00]          	|			WAVE out = FindLevelWrapper(data, level[0], edge[0], FINDLEVEL_MODE_SINGLE)
[00]          	|			break
[00]          	|		case "apfrequency":
[00]          	|			// apfrequency(data, [frequency calculation method], [spike detection crossing level])
[00]          	|			numIndices = JSON_GetArraySize(jsonID, jsonPath)
[00]          	|			ASSERT(numIndices <=3, "Maximum number of arguments exceeded.")
[00]          	|			ASSERT(numIndices >= 1, "At least one argument.")
[00]          	|
[00]          	|			WAVE data = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/0", graph = graph)
[00]          	|			if(numIndices == 3)
[00]          	|				WAVE level = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/2", graph = graph)
[00]          	|				ASSERT(DimSize(level, ROWS) == 1, "Too many input values for parameter level")
[00]          	|				ASSERT(IsNumericWave(level), "level parameter must be numeric")
[00]          	|			else
[00]          	|				Make/FREE/N=1 level = {0}
[00]          	|			endif
[00]          	|
[00]          	|			if(numIndices >= 2)
[00]          	|				WAVE method = SF_FormulaExecutor(jsonID, jsonPath = jsonPath + "/1", graph = graph)
[00]          	|				ASSERT(DimSize(method, ROWS) == 1, "Too many input values for parameter method")
[00]          	|				ASSERT(IsNumericWave(method), "method parameter must be numeric.")
[00]          	|				ASSERT(method[0] == SF_APFREQUENCY_FULL || method[0] == SF_APFREQUENCY_INSTANTANEOUS, "method parameter is invalid")
[00]          	|			else
[00]          	|				Make/FREE method = {SF_APFREQUENCY_FULL}
[00]          	|			endif
[00]          	|
[00]          	|			WAVE levels = FindLevelWrapper(data, level[0], FINDLEVEL_EDGE_INCREASING, FINDLEVEL_MODE_MULTI)
[00]          	|			variable numSets = DimSize(levels, ROWS)
[00]          	|			Make/FREE/N=(numSets) levelPerSet = str2num(GetDimLabel(levels, ROWS, p))
[00]          	|
[00]          	|			switch(method[0])
[00]          	|				case SF_APFREQUENCY_FULL:
[00]          	|					Make/N=(numSets)/D/FREE outD = levelPerSet[p] / (DimDelta(data, ROWS) * DimSize(data, ROWS))
[00]          	|					break
[00]          	|				case SF_APFREQUENCY_INSTANTANEOUS:
[00]          	|					Make/N=(numSets)/D/FREE outD
[00]          	|
[00]          	|					for(i = 0; i < numSets; i += 1)
[00]          	|						if(levelPerSet[i] <= 1)
[00]          	|							outD[i] = NaN
[00]          	|						else
[00]          	|							Make/FREE/D/N=(levelPerSet[i] - 1) distances
[00]          	|							distances[0, levelPerSet[i] - 2] = levels[i][p + 1] - levels[i][p]
[00]          	|							outD[i] = 1.0 / Mean(distances)
[00]          	|						endif
[00]          	|					endfor
[00]          	|					break
[00]          	|			endswitch
[00]          	|
[00]          	|			// @todo we assume that the x-axis of data has a ms scale
[00]          	|			outD *= 1e3
[00]          	|
[00]          	|			WAVE out = outD
[00]          	|
[00]          	|			break
[00]          	|		default:
[00]          	|			ASSERT(0, "Undefined Operation")
[00]          	|	endswitch
[00]          	|	/// @}
[00]          	|
[00]          	|	return out
[00]          	|End

*******************************************************************************************
Function: MIES_SweepFormula.ipf SF_GetSweepForFormula; Percent total 32%
*******************************************************************************************
[00]          	|static Function/WAVE SF_GetSweepForFormula(graph, rangeStart, rangeEnd, channelType, channelNumber, sweeps)
[00]          	|	String graph
[00]          	|	Variable rangeStart, rangeEnd
[00]          	|	Variable channelType, channelNumber
[00]          	|	WAVE sweeps
[00]          	|
[00]          	|	Variable i, numSweeps
[00]          	|	Variable pStart, pEnd
[00]          	|
[00]          	|	ASSERT(WindowExists(graph), "graph window does not exist")
[00]          	|	ASSERT(!IsNaN(rangeStart) && !IsNaN(rangeEnd), "Specified range not valid.")
[00]          	|	ASSERT(DimSize(sweeps, COLS) < 2, "Sweeps are one-dimensional.")
[00]          	|
[00]          	|	if(IsFinite(channelType))
[00]          	|		WAVE/T/Z traces = PA_GetTraceInfos(graph, channelType = channelType)
[00]          	|	else
[00]          	|		WAVE/T/Z traces = PA_GetTraceInfos(graph)
[00]          	|	endif
[00]          	|	if(!WaveExists(traces) || DimSize(traces, ROWS) == 0)
[00]          	|		DebugPrint("No matching sweeps for channel type: " + StringFromList(channelType, ITC_CHANNEL_NAMES))
[00]          	|		return $""
[00]          	|	endif
[00]          	|
[00]          	|	// This is a 2D-Wave sorted by sweeps and channels
[00]          	|	// It can be redimensioned to restore channel information in the "3rd" dimension.
[00]*         	|	SortColumns/A/DIML/KNDX={FindDimLabel(traces, COLS, "channelType"), FindDimLabel(traces, COLS, "channelNumber"), FindDimLabel(traces, COLS, "sweepNumber")} sortWaves=traces
[00]          	|
[00]          	|	Make/N=(DimSize(traces, ROWS))/FREE sweepListIndex
[00]          	|	for(i = 0; i < DimSize(traces, ROWS); i += 1)
[00]          	|		FindValue/V=(trunc(str2num(traces[i][%sweepNumber])))/T=(0.1) sweeps
[00]          	|		sweepListIndex[i] = V_Value != -1
[00]          	|	endfor
[00]          	|	Make/N=(DimSize(traces, ROWS))/FREE channelNumberIndex = IsFinite(channelNumber) ? str2num(traces[p][%channelNumber]) == channelNumber : 1
[00]          	|	Extract/FREE/INDX sweepListIndex, indices, (sweepListIndex[p] == 1) && (channelNumberIndex[p] == 1)
[00]          	|	numSweeps = DimSize(indices, ROWS)
[00]          	|	if(numSweeps == 0)
[00]          	|		DebugPrint("No matching sweeps")
[00]          	|		return $""
[00]          	|	endif
[00]          	|
[00]          	|	Make/FREE/T/N=(numSweeps) experiments
[00]          	|	experiments[] = traces[indices][%experiment]
[00]          	|	WAVE/Z uniqueExperiments = GetUniqueEntries(experiments)
[00]          	|	ASSERT(DimSize(uniqueExperiments, ROWS) == 1, "Sweep data is from more than one experiment. This is currently not supported.")
[00]          	|
[00]          	|	WAVE reference = $(traces[indices[0]][%fullPath])
[00]          	|	ASSERT(DimSize(reference, COLS) <= 1, "Unhandled Sweep Format.")
[00]          	|
[00]          	|	pStart = IsFinite(rangeStart) ? ScaleToIndex(reference, rangeStart, ROWS) : 0
[00]          	|	pEnd = IsFinite(rangeEnd) ? ScaleToIndex(reference, rangeEnd, ROWS) : DimSize(reference, ROWS) - 1
[00]          	|	ASSERT(pEnd < DimSize(reference, ROWS) && pStart >= 0, "Invalid sweep range.")
[01]*         	|	Make/FREE/N=(abs(pStart - pEnd), numSweeps) sweepData
[00]          	|	SetScale/P x, IndexToScale(reference, pStart, ROWS), DimDelta(reference, ROWS), sweepData
[00]          	|	for(i = 0; i < numSweeps; i += 1)
[00]          	|		WAVE sweep = $(traces[indices[i]][%fullPath])
[00]          	|		pStart = IsFinite(rangeStart) ? ScaleToIndex(sweep, rangeStart, ROWS) : 0
[00]          	|		pEnd = IsFinite(rangeEnd) ? ScaleToIndex(sweep, rangeEnd, ROWS) : DimSize(reference, ROWS) - 1
[00]          	|		ASSERT(pEnd < DimSize(sweep, ROWS) && pStart >= 0, "Invalid sweep range.")
[00]          	|		ASSERt(abs(pStart - pEnd) == DimSize(sweepData, ROWS), "Sweeps not equal.")
[00]          	|		ASSERT(DimDelta(sweep, ROWS) == DimDelta(sweepData, ROWS), "Sweeps not equal.")
[31]***       	|		sweepData[][i] = sweep[pStart + p]
[00]          	|	endfor
[00]          	|
[00]          	|	return sweepData
[00]          	|End

timjarsky commented:
@t-b using `multithread` for line 415 of SF doubles the speed on my desktop.

e.g., 
```
multithread out[indices[i]][][][] = out[indices[i]][0][0][0]

```

timjarsky commented on Packages/MIES/MIES_Utilities.ipf:
```diff
@@ -4562,23 +4574,76 @@ Function/WAVE FindLevelWrapper(data, level, edge)
 
 	ASSERT(DimSize(data, LAYERS) <= 1, "Unexpected input dimension")
+
 	Redimension/N=(numColsFixed * numRows)/E=1 data
```
@t-b I Could we copy and work with the copy instead of modifying data passed into the function?

timjarsky commented:
https://github.com/AllenInstitute/MIES/pull/384#issuecomment-559336553

@t-b I tracked this down to the fact that negative levels are stripped when setting the JSON.
![image](https://user-images.githubusercontent.com/5693697/70181041-fedc2580-1695-11ea-9e60-99a9d95136cd.png)

![image](https://user-images.githubusercontent.com/5693697/70181000-ee2baf80-1695-11ea-9f7e-a112b0da135b.png)

t-b commented:
@timjarsky Cleanuped the commits and code. Ready for action!

timjarsky commented:
@t-b this is working great and is everything I wanted. Soooo, now I'm going to add to what I want :)

Would you add a third method: APcount. It just returns the number of level crossing / number of action potentials.

t-b commented:
@timjarsky Glad you like the current state! Shouldn't we call the third method just `findlevels`?

t-b commented:
@timjarsky I've just added `apcount`.

t-b commented:
@timjarsky The test failure is real but only present on 32bit IP. I'll merge anyway but raise the required IP8 nightly once I got it fixed from WM.
